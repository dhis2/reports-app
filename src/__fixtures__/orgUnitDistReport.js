export const reportResponseData = {
    headers: [
        {
            name: 'ou',
            type: 'java.lang.String',
            hidden: false,
            meta: false,
        },
        {
            name: 'chp',
            column: 'CHP',
            type: 'java.lang.String',
            hidden: false,
            meta: false,
        },
        {
            name: 'chc',
            column: 'CHC',
            type: 'java.lang.String',
            hidden: false,
            meta: false,
        },
        {
            name: 'mchp',
            column: 'MCHP',
            type: 'java.lang.String',
            hidden: false,
            meta: false,
        },
        {
            name: 'clinic',
            column: 'Clinic',
            type: 'java.lang.String',
            hidden: false,
            meta: false,
        },
        {
            name: 'hospital',
            column: 'Hospital',
            type: 'java.lang.String',
            hidden: false,
            meta: false,
        },
    ],
    height: 16,
    rows: [
        ['Gbo', '', '1', '', '', ''],
        ['Bo', '18', '26', '58', '7', '4'],
        ['Tikonko', '1', '2', '3', '', ''],
        ['Niawa Lenga', '', '2', '1', '', ''],
        ['Bargbe', '', '1', '4', '', ''],
        ['Komboya', '', '1', '2', '', ''],
        ['Valunia', '2', '2', '3', '', ''],
        ['Lugbu', '2', '1', '6', '', ''],
        ['Selenga', '', '1', '1', '', ''],
        ['Wonde', '', '1', '2', '', ''],
        ['Bargbo', '1', '1', '5', '', ''],
        ['Kakua', '3', '7', '17', '7', '3'],
        ['Jaiama Bongor', '3', '1', '2', '', ''],
        ['Baoma', '3', '3', '7', '', ''],
        ['Bumpe Ngao', '3', '1', '4', '', '1'],
        ['Badjia', '', '1', '1', '', ''],
    ],
    width: 6,
    fileUrls: [
        {
            extension: 'xls',
            url:
                'http://localhost:8080/dhis/api/32/orgUnitAnalytics.xls?ou=O6uvpzGd5pu;YmmeuGbqOwR;JdhagCUEMbj;ARZ4y5i4reU;sxRd2XOzFbz;vWbkYPRmKyS;YuQRtpLP10I;I4jWcnFmgEC;daJPPxtIrQn;npWGUj37qDe;kU8vhUkAGaT;BGGmAwx33dj;dGheVylzol6;U6Kr7Gtpidn;KctpIIucige;zFDYIgyGmXG&ougs=J5jldMd8OHv&columns=J5jldMd8OHv',
        },
        {
            extension: 'pdf',
            url:
                'http://localhost:8080/dhis/api/32/orgUnitAnalytics.pdf?ou=O6uvpzGd5pu;YmmeuGbqOwR;JdhagCUEMbj;ARZ4y5i4reU;sxRd2XOzFbz;vWbkYPRmKyS;YuQRtpLP10I;I4jWcnFmgEC;daJPPxtIrQn;npWGUj37qDe;kU8vhUkAGaT;BGGmAwx33dj;dGheVylzol6;U6Kr7Gtpidn;KctpIIucige;zFDYIgyGmXG&ougs=J5jldMd8OHv&columns=J5jldMd8OHv',
        },
    ],
}
export const transformedTableData = {
    title: 'CHP - Bo',
    headers: [
        'Organisation Unit',
        'CHP',
        'CHC',
        'MCHP',
        'Clinic',
        'Hospital',
        'Total',
    ],
    rows: [
        ['Badjia', 0, 1, 1, 0, 0, 2],
        ['Baoma', 3, 3, 7, 0, 0, 13],
        ['Bargbe', 0, 1, 4, 0, 0, 5],
        ['Bargbo', 1, 1, 5, 0, 0, 7],
        ['Bumpe Ngao', 3, 1, 4, 0, 1, 9],
        ['Gbo', 0, 1, 0, 0, 0, 1],
        ['Jaiama Bongor', 3, 1, 2, 0, 0, 6],
        ['Kakua', 3, 7, 17, 7, 3, 37],
        ['Komboya', 0, 1, 2, 0, 0, 3],
        ['Lugbu', 2, 1, 6, 0, 0, 9],
        ['Niawa Lenga', 0, 2, 1, 0, 0, 3],
        ['Selenga', 0, 1, 1, 0, 0, 2],
        ['Tikonko', 1, 2, 3, 0, 0, 6],
        ['Valunia', 2, 2, 3, 0, 0, 7],
        ['Wonde', 0, 1, 2, 0, 0, 3],
        ['Bo', 18, 26, 58, 7, 4, 113],
    ],
}

export const chartResponseData = {
    headers: [
        {
            name: 'ou',
            type: 'java.lang.String',
            hidden: false,
            meta: false,
        },
        {
            name: 'chp',
            column: 'CHP',
            type: 'java.lang.String',
            hidden: false,
            meta: false,
        },
        {
            name: 'chc',
            column: 'CHC',
            type: 'java.lang.String',
            hidden: false,
            meta: false,
        },
        {
            name: 'mchp',
            column: 'MCHP',
            type: 'java.lang.String',
            hidden: false,
            meta: false,
        },
        {
            name: 'clinic',
            column: 'Clinic',
            type: 'java.lang.String',
            hidden: false,
            meta: false,
        },
        {
            name: 'hospital',
            column: 'Hospital',
            type: 'java.lang.String',
            hidden: false,
            meta: false,
        },
    ],
    height: 1,
    rows: [['Bo', '18', '26', '58', '7', '4']],
    width: 6,
    fileUrls: [
        {
            extension: 'xls',
            url:
                'http://localhost:8080/dhis/api/32/orgUnitAnalytics.xls?ou=O6uvpzGd5pu&ougs=J5jldMd8OHv&columns=J5jldMd8OHv',
        },
        {
            extension: 'pdf',
            url:
                'http://localhost:8080/dhis/api/32/orgUnitAnalytics.pdf?ou=O6uvpzGd5pu&ougs=J5jldMd8OHv&columns=J5jldMd8OHv',
        },
    ],
}

export const transformedChartData = {
    data: {
        labels: ['CHP', 'CHC', 'MCHP', 'Clinic', 'Hospital'],
        datasets: [
            {
                label: 'Count',
                backgroundColor: 'rgba(39, 102, 150, 0.6)',
                borderColor: 'rgb(39, 102, 150)',
                borderWidth: 1,
                hoverBackgroundColor: 'rgba(39, 102, 150, 0.8)',
                data: [18, 26, 58, 7, 4],
            },
        ],
    },
    options: {
        title: {
            display: true,
            text: 'CHP - Bo',
            fontSize: '16',
            fontFamily: "'Roboto', sans-serif",
        },
        legend: {
            position: 'bottom',
            fontFamily: "'Roboto', sans-serif",
        },
        animation: {
            duration: 180,
        },
    },
}
