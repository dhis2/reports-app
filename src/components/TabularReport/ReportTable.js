/* React */
import React, { Fragment } from 'react'
import PropTypes from 'prop-types'

export const Row = ({ cells }) => (
    <tr>
        {cells.map((cell, index) => (
            <td key={`c-${cell}-${index}`}>{cell}</td>
        ))}
    </tr>
)

Row.propTypes = {
    cells: PropTypes.arrayOf(PropTypes.string).isRequired,
}

const ReportTable = ({ content: { title, headers, rows } }) => (
    <Fragment>
        <h1>{title}</h1>
        <table>
            <thead>
                <tr className="tabular-report__header-row">
                    {headers.map((header, index) => (
                        <th key={`h-${header}-${index}`}>{header}</th>
                    ))}
                </tr>
            </thead>
            <tbody className={'tabular-report__table-body'}>
                {rows.map((cells, index) => (
                    <Row cells={cells} key={`r-${cells[0]}-${index}`} />
                ))}
            </tbody>
        </table>
    </Fragment>
)

ReportTable.propTypes = {
    content: PropTypes.object.isRequired,
}

export default ReportTable
