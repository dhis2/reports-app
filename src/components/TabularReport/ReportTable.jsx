import PropTypes from 'prop-types'
import styles from './ReportTable.module.css'

export const Row = ({ cells }) => (
    <tr className={styles.bodyRow}>
        {cells.map((cell, index) => (
            <td
                key={`c-${cell}-${index}`}
                className={`${styles.cell} ${index !== 0 ? styles.centered : ''}`}
            >
                {cell}
            </td>
        ))}
    </tr>
)

Row.propTypes = {
    cells: PropTypes.arrayOf(
        PropTypes.oneOfType([PropTypes.string, PropTypes.number])
    ).isRequired,
}

const ReportTable = ({ content: { title, headers, rows } }) => (
    <Fragment>
        <h1>{title}</h1>
        <table className={styles.table}>
            <thead>
                <tr className={styles.headerRow}>
                    {headers.map((header, index) => (
                        <th
                            key={`h-${header}-${index}`}
                            className={`${styles.headerCell} ${
                                index !== 0 ? styles.centered : ''
                            }`}
                        >
                            {header}
                        </th>
                    ))}
                </tr>
            </thead>
            <tbody>
                {rows.map((cells, index) => (
                    <Row cells={cells} key={`r-${cells[0]}-${index}`} />
                ))}
            </tbody>
        </table>
    </Fragment>
)

ReportTable.propTypes = {
    content: PropTypes.object.isRequired,
}

export default ReportTable
