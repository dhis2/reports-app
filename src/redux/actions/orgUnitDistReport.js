import { getOrgUnitDistReport } from '../../utils/api'
import {
    loadingReportDataStart,
    loadingReportDataSuccessWithFeedback,
    loadingReportDataErrorWithFeedback,
} from './reportData'

export const actionTypes = {
    SET_TABULAR_OUTPUT: 'SET_TABULAR_OUTPUT',
    SET_CHART_OUTPUT: 'SET_CHART_OUTPUT',
}
/**
 * @returns {Object}
 */
export const setOutputToChart = () => ({
    type: actionTypes.SET_CHART_OUTPUT,
})

/**
 * @returns {Object}
 */
export const setOutputToTabular = () => ({
    type: actionTypes.SET_TABULAR_OUTPUT,
})

/**
 * @returns {Function}
 */
export const loadChart = () => dispatch => {
    dispatch(setOutputToChart())
    dispatch(loadReport())
}

/**
 * @returns {Function}
 */
export const loadTable = () => dispatch => {
    dispatch(setOutputToTabular())
    dispatch(loadReport())
}

/**
 * @returns {Function}
 */
export const loadReport = () => (dispatch, getState) => {
    dispatch(loadingReportDataStart())

    const {
        organisationUnits,
        orgUnitGroupSets,
        orgUnitDistReport: { shouldShowChart },
    } = getState()
    const orgUnit = organisationUnits.selected
    const groupSetId = orgUnitGroupSets.selected

    return getOrgUnitDistReport(orgUnit, groupSetId)
        .then(response =>
            dispatch(
                loadingReportDataSuccessWithFeedback(response, shouldShowChart)
            )
        )
        .catch(error => dispatch(loadingReportDataErrorWithFeedback(error)))
}
