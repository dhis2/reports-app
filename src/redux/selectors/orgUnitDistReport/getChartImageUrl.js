import { getApi } from '../../../utils/api'

export const getChartImageUrl = state => {
    if (
        !state.organisationUnits.selected ||
        !state.orgUnitGroupSets.selected ||
        !state.orgUnitDistReport.displayImage
    ) {
        return ''
    }

    const { baseUrl } = getApi()
    const orgUnitId = state.organisationUnits.selected.id
    const orgUnitGroupSetsId = state.orgUnitGroupSets.selected
    const currentTime = new Date().getTime()

    const imageUrl = `${baseUrl}/organisationUnits/${orgUnitId}/distributionChart.png`
    const imageUrlParams = `groupSetId=${orgUnitGroupSetsId}&t=${currentTime}`

    return `${imageUrl}?${imageUrlParams}`
}
