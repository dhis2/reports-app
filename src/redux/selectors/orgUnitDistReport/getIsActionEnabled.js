export const getIsActionEnabled = state =>
    !!(
        state.organisationUnits.selected &&
        state.orgUnitGroupSets.selected &&
        !state.reportData.loading
    )
