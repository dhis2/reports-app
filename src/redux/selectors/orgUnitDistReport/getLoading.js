export const getLoading = state =>
    state.reportData.loading || state.orgUnitDistReport.chartImageLoading
