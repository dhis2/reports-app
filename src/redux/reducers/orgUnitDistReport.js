import { actionTypes } from '../actions/orgUnitDistReport'

export const defaultState = {
    shouldShowChart: false,
}

export const orgUnitDistReport = (state = defaultState, { type } = {}) => {
    switch (type) {
        case actionTypes.SET_TABULAR_OUTPUT:
            return { shouldShowChart: false }

        case actionTypes.SET_CHART_OUTPUT:
            return { shouldShowChart: true }

        default:
            return state
    }
}
