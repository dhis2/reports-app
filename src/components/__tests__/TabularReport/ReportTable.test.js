import React from 'react'
import { shallow } from 'enzyme'
import ReportTable, { Row } from '../../TabularReport/ReportTable'
import { transformedTableData } from '../../../__fixtures__/orgUnitDistReport'

describe('<ReportTable/>', () => {
    it('matches the snapshot for a given input', () => {
        const wrapper = shallow(<ReportTable content={transformedTableData} />)
        expect(wrapper).toMatchSnapshot()
    })
})

describe('<Row/>', () => {
    it('produces tr and td elements', () => {
        const cells = ['a', 'b', 'c', 'd']
        const wrapper = shallow(<Row cells={cells} />)
        expect(wrapper).toMatchSnapshot()
    })
})
