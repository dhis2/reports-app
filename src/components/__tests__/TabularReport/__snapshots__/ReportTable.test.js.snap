// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReportTable/> matches the snapshot for a given input 1`] = `
<Fragment>
  <h1>
    Facility Ownership - Bo
  </h1>
  <table>
    <thead>
      <tr
        className="tabular-report__header-row"
      >
        <th
          key="h-Organisation unit-0"
        >
          Organisation unit
        </th>
        <th
          key="h-Mission-1"
        >
          Mission
        </th>
        <th
          key="h-NGO-2"
        >
          NGO
        </th>
        <th
          key="h-Private Clinic-3"
        >
          Private Clinic
        </th>
        <th
          key="h-Public facilities-4"
        >
          Public facilities
        </th>
        <th
          key="h-Total-5"
        >
          Total
        </th>
      </tr>
    </thead>
    <tbody
      className="tabular-report__table-body"
    >
      <Component
        cells={
          Array [
            "Badjia",
            "0",
            "0",
            "0",
            "2",
            "2",
          ]
        }
        key="r-Badjia-0"
      />
      <Component
        cells={
          Array [
            "Baoma",
            "0",
            "0",
            "0",
            "13",
            "13",
          ]
        }
        key="r-Baoma-1"
      />
      <Component
        cells={
          Array [
            "Bargbe",
            "0",
            "0",
            "0",
            "5",
            "5",
          ]
        }
        key="r-Bargbe-2"
      />
      <Component
        cells={
          Array [
            "Bargbo",
            "0",
            "0",
            "0",
            "7",
            "7",
          ]
        }
        key="r-Bargbo-3"
      />
      <Component
        cells={
          Array [
            "Bumpe Ngao",
            "1",
            "0",
            "0",
            "9",
            "10",
          ]
        }
        key="r-Bumpe Ngao-4"
      />
      <Component
        cells={
          Array [
            "Gbo",
            "0",
            "0",
            "0",
            "1",
            "1",
          ]
        }
        key="r-Gbo-5"
      />
      <Component
        cells={
          Array [
            "Jaiama Bongor",
            "0",
            "0",
            "0",
            "6",
            "6",
          ]
        }
        key="r-Jaiama Bongor-6"
      />
      <Component
        cells={
          Array [
            "Kakua",
            "5",
            "3",
            "11",
            "19",
            "38",
          ]
        }
        key="r-Kakua-7"
      />
      <Component
        cells={
          Array [
            "Komboya",
            "0",
            "0",
            "0",
            "3",
            "3",
          ]
        }
        key="r-Komboya-8"
      />
      <Component
        cells={
          Array [
            "Lugbu",
            "0",
            "0",
            "0",
            "9",
            "9",
          ]
        }
        key="r-Lugbu-9"
      />
      <Component
        cells={
          Array [
            "Niawa Lenga",
            "0",
            "0",
            "0",
            "3",
            "3",
          ]
        }
        key="r-Niawa Lenga-10"
      />
      <Component
        cells={
          Array [
            "Selenga",
            "0",
            "0",
            "0",
            "2",
            "2",
          ]
        }
        key="r-Selenga-11"
      />
      <Component
        cells={
          Array [
            "Tikonko",
            "0",
            "0",
            "0",
            "6",
            "6",
          ]
        }
        key="r-Tikonko-12"
      />
      <Component
        cells={
          Array [
            "Valunia",
            "0",
            "0",
            "0",
            "7",
            "7",
          ]
        }
        key="r-Valunia-13"
      />
      <Component
        cells={
          Array [
            "Wonde",
            "0",
            "0",
            "0",
            "3",
            "3",
          ]
        }
        key="r-Wonde-14"
      />
      <Component
        cells={
          Array [
            "Bo",
            "6",
            "3",
            "11",
            "95",
            "115",
          ]
        }
        key="r-Bo-15"
      />
    </tbody>
  </table>
</Fragment>
`;

exports[`<Row/> produces tr and td elements 1`] = `
<tr>
  <td
    key="c-a-0"
  >
    a
  </td>
  <td
    key="c-b-1"
  >
    b
  </td>
  <td
    key="c-c-2"
  >
    c
  </td>
  <td
    key="c-d-3"
  >
    d
  </td>
</tr>
`;
